syntax = "proto3";

import "google/protobuf/timestamp.proto";

/* LocalToAnyFollow represents either a local user following another local
 * user, or a local user following a foreign user. The fact that these
 * separate concepts are shared by one type allows the frontend to send the
 * same request to the backend without parsing the username of the followed
 * account (eg. to find out if it is local), or needing to verify that it
 * exists. */
message LocalToAnyFollow {
    /* The handle of the local user who has requested to follow someone, eg.
     * "admin". */
    string follower = 1; 
    /* The full username of the person being followed, eg. "admin", "@admin",
     * "admin@rabble.net", or "@aaron@devoxel.net". */
    string followed = 2;
    google.protobuf.Timestamp datetime = 3;
}

/* ForeignToLocalFollow represents a user on a different instance requesting
 * to follow a user on this instance. */
message ForeignToLocalFollow {
    /* The fully-qualified username of the foreign user who has requested to
     * follow a user on this instance. Eg. "@aaron@devoxel.net". */
    string follower = 1;
    /* The handle of the user on this instance that has been followed, eg.
     * "admin". */
    string followed = 2;
    google.protobuf.Timestamp datetime = 3;
}

/* TODO(iandioch): Consider merging with DbFollowResponse in
 * services/database/database.proto */
message FollowResponse {
    enum ResultType {
        OK = 0;
        ERROR = 1;
        // More types can be added here.
    }

    ResultType result_type = 1;

    /* Should only be set if result_type is not OK. */
    string error = 2;
}

service Follows {
  rpc SendFollowRequest(LocalToAnyFollow) returns (FollowResponse);
  rpc ReceiveFollowRequest(ForeignToLocalFollow) returns (FollowResponse);
}
